File: Java/State/CellPhone/src/com/ericgtkb/CellPhone.java
package com.ericgtkb;

public class CellPhone {
    private String name;

    // States are often singletons. Here for simplicity they are just normal objects.
    private CellPhoneMode normalMode;
    private CellPhoneMode silentMode;
    private CellPhoneMode airplaneMode;

    private CellPhoneMode mode;

    public CellPhone(String name) {
        this.name = name;
        normalMode = new NormalMode(this);
        silentMode = new SilentMode(this);
        airplaneMode = new AirplaneMode(this);
        mode = normalMode;
    }

    public String getName() {
        return name;
    }

    // Methods for changing modes (states). This is sometimes delegated to the states.
    public void goNormal() {
        mode = normalMode;
        System.out.println("Switching to normal mode...");
    }
    public void goSilent() {
        mode = silentMode;
        System.out.println("Switching to silent mode...");
    }

    public void goAirplane() {
        mode = airplaneMode;
        System.out.println("Switching to airplane mode...");
    }

    public void newMessage() {
        mode.newMessage();
    }
    public void newPhoneCall(){
        mode.newPhoneCall();
    }
    public void newAppNotification() {
        mode.newAppNotification();
    }
}

File: C++/State/CellPhone/cell_phone.cc
#include "cell_phone.h"
#include "cell_phone_mode.h"

CellPhone::CellPhone(const std::string &name) : name(name),
                                                normal_mode{std::make_unique<NormalMode>(this)},
                                                silent_mode{std::make_unique<SilentMode>(this)},
                                                airplane_mode{std::make_unique<AirplaneMode>(this)} {
    mode = normal_mode.get();
}

const std::string &CellPhone::get_name() const {
    return name;
}

void CellPhone::go_normal() {
    mode = normal_mode.get();
}

void CellPhone::go_silent() {
    mode = silent_mode.get();
}

void CellPhone::go_airplane() {
    mode = airplane_mode.get();
}

void CellPhone::new_message() {
    mode->new_message();
}

void CellPhone::new_phone_call() {
    mode->new_phone_call();
}

void CellPhone::new_app_notification() {
    mode->new_app_notification();
}

